_posts/2017-10-04-Law-of-quality.md

```diff
@@ -0,0 +1,106 @@
+
+Okay, but surely the situation is better in SQL databases? We'll define our table like this:
+```
+CREATE TABLE dadas
+  (
+    le_time TIMESTAMPT NOT NULL
+  ) ;
+```
+
+Oops, we just failed. According to [PostgreSQL documentation](https://www.postgresql.org/docs/9.1/static/datatype-datetime.html) this leaves out timezone, which means effectively system locale. 
+
+```
+The SQL standard requires that writing just timestamp be equivalent to timestamp without 
+time zone, and PostgreSQL honors that behavior. (Releases prior to 7.3 treated it as 
+timestamp with time zone.)
+```
+
+It is inadvisable to assume that system default locale in the database server is the same as in the application server, unless you somehow explicitly control it. But let's be gracious and assume you have everything under control in the backend. What about the user interface? Enter Javascript and let the mortals tremble.
+
+```
+new Date(2012,12,12)
+ > Sat Jan 12 2013 00:00:00 GMT+0200 (EET)
+new Date(2012,0,1)
+ > Sun Jan 01 2012 00:00:00 GMT+0200 (EET)
+new Date(2012,-1,1)
+ > Thu Dec 01 2011 00:00:00 GMT+0200 (EET)
+```
+
+Javascript is kind enough to allow data slip without warnings or exceptions like that. You get dates, just maybe a bit different dates from what you expected. And what might be the timezone should you do something like that? You can't rely on the user's browser to have any specific timezone set even if you can control the backend server. 
+
+Good luck getting all these layers to work perfectly in all situations.
+
+![Die you computer](/img/destroy-computer.jpg)
+
+## The characters
+
+Unless you are a veteran with a lot of scars, this [old article about encoding](https://www.joelonsoftware.com/2003/10/08/the-absolute-minimum-every-software-developer-absolutely-positively-must-know-about-unicode-and-character-sets-no-excuses/) from Joel Spolsky should give you something to think about.
+
+As with dates, it doesn't end here. With Java your source code and strings at least have some defined encodings and characters sets. Python developers are not so fortunate, see this: [Unicode in Python](https://docs.python.org/2/howto/unicode.html). 
+
+Again, remember to check your relational database for surprises. On particularly interesting one is the order of result set we get from this innocent SQL statement:
+
+```
+select * from winners order by name;
+```
+
+As article on [Alphabetical Order](https://en.wikipedia.org/wiki/Alphabetical_order) in Wikipedia informs us, order of letters in the alphabet depends on the country. V may come before W, or can it? And the order has been changed in some countries quite recently. This may result in a bit surprising orderings as some characters can even have equal standing in the order. Do you actually know what ordering your database server is using now? Did you specify it?
```

Comment:
“As **the** Wikipedia article on… informs us, **the** order of **the** letters…”

“in orderings that are a bit surprising”